pkg/discovery/grpcresolver/grpc_resolver.go

```diff
@@ -0,0 +1,155 @@
+	"google.golang.org/grpc/resolver"
+
+	"github.com/jaegertracing/jaeger/pkg/discovery"
+)
+
+// Resolver uses notifier to fetch list of available hosts
+type Resolver struct {
+	scheme            string
+	cc                resolver.ClientConn
+	notifier          discovery.Notifier
+	discoverer        discovery.Discoverer
+	logger            *zap.Logger
+	discoCh           chan []string // used to receive notifications
+	discoveryMinPeers int
+	mu                sync.Mutex
+	salt              []byte
+	hasher            hash.Hash32
+	wg                sync.WaitGroup
+}
+type hostScore struct {
+	address string
+	score   uint32
+}
+
+type hostScores []hostScore
+
+func (s hostScores) Len() int           { return len(s) }
+func (s hostScores) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+func (s hostScores) Less(i, j int) bool { return s[i].score < s[j].score }
+
+// New intialize a new grpc resolver with notifier
+func New(
+	notifier discovery.Notifier,
+	discoverer discovery.Discoverer,
+	logger *zap.Logger,
+	discoveryMinPeers int,
+) *Resolver {
+	seed := time.Now().UnixNano()
+	random := rand.New(rand.NewSource(seed))
+	r := &Resolver{
+		notifier:          notifier,
+		discoverer:        discoverer,
+		discoCh:           make(chan []string, 100),
+		logger:            logger,
+		discoveryMinPeers: discoveryMinPeers,
+		salt:              []byte(strconv.FormatInt(random.Int63(), 10)), // random salt for rendezvousHash
+		scheme:            strconv.FormatInt(seed, 36),                   // make random scheme which will be used when registering
+		hasher:            fnv.New32(),
+	}
+	// TODO not sure if there's an equivalent way for grpc to maintain connection like what tchannel did?
+
+	// Register the resolver with grpc so it's available for grpc.Dial
+	resolver.Register(r)
+
+	// Register the discoCh channel with notifier so it continues to fetch a list of host/port
+	notifier.Register(r.discoCh)
+	return r
+}
+
+// Build returns itself for Resolver, because it's both a builder and a resolver.
+func (r *Resolver) Build(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOption) (resolver.Resolver, error) {
+	r.cc = cc
+
+	// Update conn states if proactively updates already work
+	instances, err := r.discoverer.Instances()
+	if err != nil {
+		return nil, err
+	}
+	r.cc.UpdateState(resolver.State{Addresses: generateAddresses(instances)})
+	r.wg.Add(1)
+	go r.watcher()
+	return r, nil
+}
+
+// Scheme returns the test scheme.
+func (r *Resolver) Scheme() string {
+	return r.scheme
+}
+
+// ResolveNow is a noop for Resolver.
+func (r *Resolver) ResolveNow(o resolver.ResolveNowOption) {}
+
+func (r *Resolver) watcher() {
+	defer r.wg.Done()
+	for latestHostPorts := range r.discoCh {
+		r.mu.Lock()
+		r.logger.Info("Received updates from notifier", zap.Strings("hostPorts", latestHostPorts))
+		r.cc.UpdateState(resolver.State{Addresses: generateAddresses(rendezvousHash(latestHostPorts, r.salt, r.hasher, r.discoveryMinPeers))})
```

Comment:
- pull generateAddresses() call into  a variable
- if you make it a member function, you don't need to pass parameters that are already accessible from `r.`

Reply:
Regarding using the `grpc.Roundrobin` and `grpc.WithBalancer`, I believe those were deprecated. 

![image](https://user-images.githubusercontent.com/7311035/56901111-5e6db500-6a65-11e9-9548-81fbcc418803.png)

![image](https://user-images.githubusercontent.com/7311035/56901178-8a893600-6a65-11e9-90a0-d91a9598f4d7.png)